source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,154 @@
+
+- When a VM stops for any reason on a connected host, we store the latest state of the data.
+- When a host or a storage get disconnected, we store the latest known state of the data.
+- When a VM is restarted without Engine assistance, it uses the final data from the last run.
+- We don't store inconsistent data.  swtpm apparently [doesn't store changes atomically](https://github.com/stefanberger/swtpm/blob/609dfd873a31fc0402752c50d1d5cfa58949507f/src/swtpm/swtpm_nvfile.c#L492), so whether we copy swtpm data or let swtpm write to shared storage, we are at risk of race with swtpm writes on copying or when the storage gets suddenly disconnected.
+- Temporary network disconnections, lost events sent from Vdsm to Engine or temporary failures when delivering or processing Vdsm responses to Engine don't cause permanent loss of data updates.
+
+The options to deliver updated TPM data are similar to obtaining TPM data on VM start, with the important distinction that this time they are write instead of read operations, with the risk of data loss.  We can:
+
+- Send the updated data to Engine in events, current API call responses (stats, destroy) or using a new Vdsm API call for the given purpose.
+- Store the data in metadata, to be read and processed in Engine monitoring.
+- Store the data to shared storage from Vdsm.
+- Let swtpm store the data directly to the shared storage.
+- Some combination of the options above.
+
+Different options are discussed in the following subsections.  We can probably assume that TPM data updates are not going to be very frequent.  We mustn't forget about reboots, snapshots, imports and exports.
+
+#### Using API calls
+
+TPM data can be stored in the Engine database and transferred in both the directions using Vdsm API calls.  It can be provided to the VM as an additional parameter to VM.create call.  Data changes can be reported back to Engine using:
+
+- Events sent from Vdsm.
+- As a return value parameter of VM.destroy call.
+- As an additional value in VM stats.
+- On a request from Engine using a newly introduced Vdsm API call.
+
+TPM data can be copied to swtpm data location before the VM is created.  Once VM is started, updated TPM data can be read from the swtpm data location.  A problem is how to read data from swtpm consistently, avoiding collisions with swtpm writes.  Considering TPM data is small, written locally and not updated frequently, a good enough approach should be to have a periodic Vdsm job watching for swtpm data changes.  If it detects a data change, it keeps watching the data and once it detects the same data in two consecutive job runs, it can report it as changed data relatively safely (unless there is some fatal condition such as a full local disk).  This mechanism would be common to all the approaches except for shared storage directly used by swtpm.
```

Comment:
This is not so simple, since the data may change again after we detected a change, but it can work.